### DISCUSSION TITLE
New collections

### DISCUSSION THREAD

**Discussion by iampoundcake** [chrismwilliams/astro-theme-cactus#248] [Q&A]
Anyone else having difficulty with creating new collections? I will absolutely concede that it is a user-error/user ignorance. My buddy (who codes) and i were doing some root cause and testing, and we kept hitting a wall when we finally go to inject a collection in a page. I know other astro sites are simpler and less conducive to scale, so I wanted to know, just to make sure, how I should go about new collections. I also tried chatgpt but i feel like it's just breaking more things.

1. this theme uses the src/data/post.ts for injection right? does a new data file need to be created per collection?
I did try to do this with a test file, but i keep getting an error that the "map" is unrecognized or undefined. The original code says "allPosts..." and works. (i used "TEST" as an example of the definition of the collection)
`allTESTByDate.map((p) => (`

Anyway, any help or thoughts are appreciated. I could just use the tags, and figure out a way to stylize pages based on tags, but i'd like to figure this out if i can.

**Comment by chrismwilliams** [comment]
Hi @iampoundcake. I hope I haven't made it more difficult, my intention was to add example(s) for people to _copy_, so apologies if that's not the case.

To answer your question, the theme uses `src/data/posts.ts` as a collection of helper functions, mainly for consistency when retrieving the 'post' content collection, so you don't need to follow this format, but they are specific to the "post" collection. 

It sounds like you want to add another collection, which does mean you will have to define [multiple](https://docs.astro.build/en/guides/content-collections/#defining-multiple-collections). Each collection is separate from each other and you'll need to use Astro's [getCollection](https://docs.astro.build/en/guides/content-collections/#filtering-collection-queries) to query each collection you add separately.

Let me know if any of that doesn't makes sense, if your repo is public I could take a quick look if required.

Hope that helps.

**Reply by iampoundcake** [reply]
Thank you so much, @chrismwilliams ! 

And I don't think you've made it more difficult at all. I think, if anything, you built something the proper way with the ability to scale (which is great!). I think I'm just getting lost at what files i need to edit and/or add to accommodate a new collection, if at all. 

So it sounds like the "post.ts" file equivalent isn't necessary for each collection. I'll try on my end today and follow up.

thanks again!!

**Reply by iampoundcake** [reply]
So I created a new test branch based on what I am working on. I've now gotten the new collection "Dailies" to appear on a src/pages/dailies/index.astro, but now the .md file is 404'ing. It's very barebones. When I try to apply similar code as the posts/blog for tagging or other features, I usually break many other things and get stuck at the "undefined 'map'" Anyway, I can try to do that later once I get this simple set up working. 

https://github.com/iampoundcake/joe-website/tree/codespace-probable-acorn-9gj9qq4x95wh459

**Reply by chrismwilliams** [reply]
Okay, I think I can see the issue. I guess the first thing to decide, you may have already but just so I'm clear, is whether you want your "Dailies" content collection to be the same shape (in terms of properties) as a "Post". ATM they're exactly the same, which isn't a problem, but they don't _need_ to be. 

Depending on the above will affect reusing a lot of "Post" components, as they currently expect a `CollectionEntry<"post">` type, which would need to change if you want to reuse it for dailies. 

To quickly summarise the main issue with 404 though:
 - an errand curly bracket in `src/pages/dailies/[...slug}.astro`. You also probably don't want the rest parameter either as it's a [catch-all](https://docs.astro.build/en/guides/routing/#rest-parameters). Something like `src/pages/dailies/[slug].astro` would work.
 - the href in [this link](https://github.com/iampoundcake/joe-website/blob/c91a857f5768f3ae971e6081819b8102a22088de/src/pages/dailies/index.astro#L41) should be ``href={`/dailies/${p.slug}/`}``
 - the [props in getStaticPaths](https://github.com/iampoundcake/joe-website/blob/c91a857f5768f3ae971e6081819b8102a22088de/src/pages/dailies/%5B...slug%7D.astro#L13) should be `props: { daily },`
 - following the above, the [daily const](https://github.com/iampoundcake/joe-website/blob/c91a857f5768f3ae971e6081819b8102a22088de/src/pages/dailies/%5B...slug%7D.astro#L19) should be `const { daily } = Astro.props;`

That hopefully should fix generating the correct links and slugs/pages.

**Reply by iampoundcake** [reply]
I am humbled by your assistance man! Thank you so much. I'm gonna give it a whirl today. I shout you out in my first post from a few weeks ago and I'm glad I did bc you are the man. 

Much appreciated 🙏🙏

**Reply by iampoundcake** [reply]
I was finally able to get the 404 to work and pull the post: https://probable-acorn-9gj9qq4x95wh459-4321.app.github.dev/dailies/2024-04-07/

Now I am trying to figure out how to ensure the markdown translates to the post because it's registering as plain. (please feel free to ignore if this becomes annoying lol)

**Reply by chrismwilliams** [reply]
If you're talking about it being unstyled (no css), that's likely because the `<Content>` should be wrapped by an element with the tailwindcss class of "prose". If that's not what you mean, [Content](https://docs.astro.build/en/reference/api-reference/#content-) represents the body of the markdown file, so there could be a syntax issue in the associated .md/.mdx file.

I also forgot to add above that from [L29](https://github.com/iampoundcake/joe-website/blob/c91a857f5768f3ae971e6081819b8102a22088de/src/pages/dailies/%5B...slug%7D.astro#L29) to [L43](https://github.com/iampoundcake/joe-website/blob/c91a857f5768f3ae971e6081819b8102a22088de/src/pages/dailies/%5B...slug%7D.astro#L43) is an issue. `dailies.data.toc` doesn't exist on your schema, so it will always be false.

**Reply by iampoundcake** [reply]
Yup, that totally makes sense. where can i donate to your coffee fund??
